aes/src/x86.rs

```diff
@@ -0,0 +1,728 @@
+                unsafe {
+                    let simd_256_keys = self
+                        .simd_256_keys
+                        .get_or_init(|| vaes256::encdec::broadcast_keys(&self.keys));
+                    vaes256::encdec::encrypt30(simd_256_keys, blocks);
+                }
+            }
+            #[inline]
+            fn encrypt_tail_blocks(&self, blocks: InOutBuf<'_, '_, Block>) {
+                let backend = self;
+
+                let mut rem = blocks.len();
+                let (mut iptr, mut optr) = blocks.into_raw();
+
+                let backend = $name_backend::Ni::from(backend);
+                while rem >= backend.par_blocks() {
+                    let blocks = unsafe { InOut::from_raw(iptr.cast(), optr.cast()) };
+                    backend.encrypt_par_blocks(blocks);
+                    rem -= backend.par_blocks();
+                    iptr = unsafe { iptr.add(backend.par_blocks()) };
+                    optr = unsafe { optr.add(backend.par_blocks()) };
+                }
+
+                while rem > 0 {
+                    let block = unsafe { InOut::from_raw(iptr, optr) };
+                    backend.encrypt_block(block);
+                    rem -= 1;
+                    iptr = unsafe { iptr.add(1) };
+                    optr = unsafe { optr.add(1) };
+                }
+            }
+        }
+        #[cfg(all(target_arch = "x86_64", aes_avx512))]
+        impl<'a> BlockCipherEncBackend for $name_backend::Vaes512<'a> {
+            #[inline]
+            fn encrypt_block(&self, block: InOut<'_, '_, Block>) {
+                unsafe {
+                    self::ni::encdec::encrypt(self.keys, block);
+                }
+            }
+            #[inline]
+            fn encrypt_par_blocks(&self, blocks: InOut<'_, '_, cipher::ParBlocks<Self>>) {
+                unsafe {
+                    let simd_512_keys = self
+                        .simd_512_keys
+                        .get_or_init(|| vaes512::encdec::broadcast_keys(&self.keys));
+                    vaes512::encdec::encrypt64(simd_512_keys, blocks);
+                }
+            }
+            #[inline]
+            fn encrypt_tail_blocks(&self, blocks: InOutBuf<'_, '_, Block>) {
+                let backend = self;
+
+                let mut rem = blocks.len();
+                let (mut iptr, mut optr) = blocks.into_raw();
+
+                let backend = &$name_backend::Vaes256::from(backend);
+                if backend.features.has_vaes256() {
+                    while rem >= backend.par_blocks() {
+                        let blocks = unsafe { InOut::from_raw(iptr.cast(), optr.cast()) };
+                        backend.encrypt_par_blocks(blocks);
+                        rem -= backend.par_blocks();
+                        iptr = unsafe { iptr.add(backend.par_blocks()) };
+                        optr = unsafe { optr.add(backend.par_blocks()) };
```

Comment:
I would prefer to write this code using safe code. See the `InOutBuf::into_chunks` method.